File: 09/python/main.py
import math


def neighbors(x, y):
    def in_bounds(x, y):
        return 0 <= x < len(grid) and 0 <= y < len(grid[0])

    return ((x+dx, y+dy)
            for dx, dy in ((0, 1), (0, -1), (1, 0), (-1, 0))
            if in_bounds(x+dx, y+dy))


def part1(grid):
    def is_low_point(x, y):
        return all(grid[x][y] < grid[_x][_y]for _x, _y in neighbors(x, y))

    return sum((
        grid[x][y] + 1
        for x in range(len(grid))
        for y in range(len(grid[0]))
        if is_low_point(x, y)
    ))


def part2(grid):
    def explore(x, y):
        if grid[x][y] == 9:
            return 0
        grid[x][y] = 9
        return 1 + sum(explore(*n) for n in neighbors(x, y))

    basin_sizes = []
    for x in range(len(grid)):
        for y in range(len(grid[x])):
            if grid[x][y] != 9:
                basin_sizes.append(explore(x, y))
    basin_sizes.sort(reverse=True)
    return math.prod(basin_sizes[:3])


with open("../input.txt") as f:
    grid = [[int(c) for c in line.strip()] for line in f]
    # print(grid)
    print(part1(grid))
    print(part2(grid))

File: 09/rust/src/main.rs
use itertools::Itertools;

fn main() {
    let mut grid = include_str!("../../input.txt")
        .lines()
        .map(|line| {
            line.chars()
                .map(|c| c.to_digit(10).expect("illegal digit"))
                .collect()
        })
        .collect::<Vec<Vec<u32>>>();
    println!("{}", part1(&grid));
    println!("{}", part2(&mut grid));
}

fn part1(grid: &[Vec<u32>]) -> u32 {
    (0..grid.len())
        .cartesian_product(0..grid[0].len())
        .filter(|(x, y)| {
            neighbors(*x, *y, grid)
                .into_iter()
                .all(|(x_, y_)| grid[*x][*y] < grid[x_][y_])
        })
        .map(|(x, y)| grid[x][y] + 1)
        .sum()
}

fn part2(grid: &mut [Vec<u32>]) -> u32 {
    let mut basin_sizes = Vec::new();
    for x in 0..grid.len() {
        for y in 0..grid[x].len() {
            if grid[x][y] != 9 {
                basin_sizes.push(explore(grid, x, y));
            }
        }
    }
    basin_sizes.sort_unstable();
    basin_sizes.iter().rev().take(3).product()
}

fn explore(grid: &mut [Vec<u32>], x: usize, y: usize) -> u32 {
    if grid[x][y] == 9 {
        0
    } else {
        grid[x][y] = 9;
        1 + neighbors(x, y, grid)
            .iter()
            .map(|(x_, y_)| explore(grid, *x_, *y_))
            .sum::<u32>()
    }
}

fn neighbors(x: usize, y: usize, grid: &[Vec<u32>]) -> Vec<(usize, usize)> {
    [(0, 1), (0, -1), (1, 0), (-1, 0)]
        .iter()
        .map(move |(dx, dy)| (x as i32 + dx, y as i32 + dy))
        .filter(|(x, y)| {
            0 <= *x && (*x as usize) < grid.len() && 0 <= *y && (*y as usize) < grid[0].len()
        })
        .map(|(x, y)| (x as usize, y as usize))
        .collect()
}
